부산대 FE 손정훈 - 2단계 상품 상세 페이지 API 구현하기

SonJH7 opened the pull request:
## 구현사항
 - 상품 상세 페이지 UI 전면 수정 및 탭 메뉴, 하단 고정 바 추가로 기능을 강화했습니다.
 - React Query 기반의 API 훅을 확장하여 상품 상세 정보, 후기, 위시 데이터를 가져오고 좋아요 수를 낙관적 업데이트로 처리합니다.
 - 공통 Query Key 정의를 추가해 캐시 키 관리를 일관되게 했습니다.
 - ErrorBoundary 컴포넌트를 도입해 렌더링 오류 시 대체 화면을 표시합니다.

## 구현 간 에러사항
 - 프록시 설정 문제
외부 API 호출 과정에서 프록시 에러가 발생하여 fix(proxyerror) 커밋으로 해결했습니다.

- API 응답 구조 문제
상품 상세 정보 API에서 announcement 배열이 없는 경우 오류가 발생해, 존재할 때만 데이터를 출력하도록 수정했습니다.

 - 상세 탭 로직 오류
상세 정보 탭에서 필요한 데이터가 제대로 표시되지 않는 문제가 있어, 로직을 수정하여 정상 출력되도록 했습니다.

SonJH7 committed:
feat(gitattributes): eol 관련 개행 일치

SonJH7 committed:
refactor(querykeys): querykeys들을 상수로 분리해 정의하고 product, theme API 가각에서 이용하도록 변경

SonJH7 committed:
Update ReadME.md

SonJH7 committed:
feat(productdetailpage): 페이지 구축 및 컴포넌트 들 적용, querykey 상수화

SonJH7 committed:
feat(apiproduct): api 구축하고 경로 등록을 하였습니다

SonJH7 committed:
feat(errorboundary): 에러 발생 시 대체 화면을 보여 주는 ErrorBoundary 컴포넌트 구현

SonJH7 committed:
fix(proxyerror): 미션 구현 중 발생한 proxy 에러 해결

SonJH7 committed:
fix(productapi): 상품 상세 정보에 announcement 배열이 존재할 때만 목록을 출력 하도록 변경하여 오류 방지, react query 훅의 반환 타입 수정

SonJH7 committed:
design(productdetailpage): UI 디자인  전면 수정, 탭 메뉴와 하단 고정바를 도입 및 버튼들을 위한 스타일 새롭게 정의

SonJH7 committed:
fiX(detaildata): 상세 정보 탭에서의 로직 수정으로 출력가능하게함

SonJH7 committed:
refactor(all): 코드 리팩토링

- useHighlightReviewQuery 함수의 제너릭 타입 수정
- 안 쓰는 hook 제거

ddongule commented on src/api/client.ts:
```diff
@@ -1,5 +1,6 @@
 import type { QueryFunctionContext } from '@tanstack/react-query'
 
+const responseCache = new Map<string, unknown>()
```
responseCache는 선언만 되어있고 실제로 사용되지 않았네욥! 어떤 이유로 만들어두신걸까요?

SonJH7 replied:
React Query 자체의 캐싱 기능이 잘 적용되어 있지 않는거 같아 혹시나 수동 캐싱으로 해봤는데, 컴퓨터를 껐다 켜니 react query 만으로 해결이 되어 step3 에서는 뺐습니다!

ddongule commented on src/api/client.ts:
```diff
@@ -1,5 +1,6 @@
 
+const responseCache = new Map<string, unknown>()
 export interface FetchApiOptions {
   method?: string
```
(이번 diff는 아니지만 하핫) method는 'GET', 'POST' 등으로 좁혀줘도 좋을 것 같네욥!

SonJH7 replied:
넵 한 번 적용해보겠습니다! 처음 고려하는 사항이라 좋은 피드백 감사합니다!

ddongule commented on src/api/client.ts:
```diff
@@ -34,16 +35,32 @@ export async function fetchApi<T>(
+    cache: 'no-store',
   })
-  const json = await res.json()
+  const text = await res.text()
+  let json: any
+  try {
+    json = text ? JSON.parse(text) : undefined
+  } catch {
+    json = undefined
+  }
```
응답이 JSON인지 파악하는 용도군욥! 👍 
이렇게 할 수도 있지만 좀 더 단순한 방법은 headers 안의 Content-Type이 application/json인지 체크하는 방법도 있겠어용!

ddongule commented on src/api/client.ts:
```diff
@@ -34,16 +35,32 @@ export async function fetchApi<T>(
+      json?.data?.statusCode || json?.statusCode || res.status
+    throw error
+  }
+    if (res.status === 304) {
+    return undefined as T
+  }
+    if (json === undefined) {
+    return undefined as T
+  }
```
undefined를 반환하면서 리턴타입을 T로 해주신 이유가 있나요? 🧐

SonJH7 replied:
304 Not Modified처럼 응답 본문이 없는 경우를 처리하기 위해 undefined를 반환하지만, 함수 시그니처가 Promise<T>로 고정돼 있어 타입 오류를 피하려고 undefined as T로 했는데, 알아보니 이 방식은 타입 안전성이 떨어지는 것 같습니다.  함수 반환 타입을 Promise<T | undefined>처럼 명시적으로 표현하거나 304 응답을 별도로 처리하는 쪽으로 표현 하면 될까요?

ddongule replied:
말씀주신대로 요렇게 코드를 만들게되면 안정성이 떨어질 수 있어서, 말씀주신 방법들 중에서는 에러마다 따로 처리하는 것보다는 Promise<T | undefined> 처럼 명시적으로 표현해두고 사용하는 쪽에서 data가 undefined인지 한번 체킹해서 UI를 그리는 방식이 더 좋을 것 같아요! 
다른 방법으로는 응답 값을 래핑해 사용하는 ApiResponse<T> 요렇게 타입을 정의해 사용하는 방법도 있으니 한번 찾아보셔두 좋을 것 같구요!

ddongule commented on src/api/client.ts:
```diff
@@ -34,16 +35,32 @@ export async function fetchApi<T>(
+  if (!res.ok) {
+  const message =
+      json?.data?.message || json?.message || `Invalid response from ${url}`
+  const error = new Error(message)
+    ;(error as any).statusCode =
+      json?.data?.statusCode || json?.statusCode || res.status
+    throw error
+  }
```
prettier가 제대로 동작하지 않는 것 같아 한번 살펴주세용!

SonJH7 replied:
npm run format으로 한번 정돈 하는 걸 까먹고 올렸습니다! 잘 적용해서 올리겠습니다!

ddongule replied:
settings에 format on save 기능도 있으니 한번 고 기능도 활용해보시는걸 추천드려용!

SonJH7 replied:
넵 감사합니다!

ddongule commented on src/api/product.ts:
```diff
@@ -64,9 +73,106 @@ export function useProductSummaryQuery(
   return useQuery<ProductSummary, Error>({
-    queryKey: ['products', productId, 'summary'],
-    queryFn: () => fetchProductSummary(productId!),
+    queryKey: productKeys.summary(productId!),
+    queryFn: () => fetchProductSummary(productId!),    enabled: productId !== undefined,
```
non null assertion 을 사용하는 것보다는 조건을 붙여주는 편이 좋겠어요!

SonJH7 replied:
넵 조건을 붙혀서 다시 리뷰하겠습니다!

ddongule commented on src/api/product.ts:
```diff
@@ -64,9 +73,106 @@ export function useProductSummaryQuery(
+  return fetchApi<WishInfo>(`/api/products/${productId}/wish`)
+}
+
+export async function postWish(productId: number): Promise<void> {
+  await fetchApi(`/api/products/${productId}/wish`, { method: 'POST' })
+}
```
return 값이 없어도 괜찮은 부분인가요?

SonJH7 replied:
현재 postWish는 서버에 위시 요청만 보내고 응답 데이터는 사용하지 않는 것 같아 Promise<void>로 선언해도 문제 없는 것 같습니다. 호출하는 쪽(useWishMutation)에서도 반환 값을 필요로 하지 않는 걸로 확인 했습니다!

ddongule commented on src/api/themes.ts:
```diff
@@ -61,8 +61,7 @@ export function useThemeProductsQuery(
   return useQuery<ThemeProductsResponse, Error>({
-    queryKey: ['themes', themeId, 'products', params],
-    queryFn: () => fetchThemeProducts(themeId!, params),
+    queryKey: themeKeys.products(themeId!, params),    queryFn: () => fetchThemeProducts(themeId!, params),
```
여기두 같은리뷰입니당!

SonJH7 replied:
반환 값이 없어도 괜찮은 부분과 같은 리뷰라는 얘기인지 여쭤봐도 될까요?

ddongule replied:
앗 제가 너무 모호하게 작성했네요 ㅋㅋㅋㅋ 띄어쓰기가 이상해서 요거도 prettier 이슈가 있는 것 같다는 리뷰였습니다 죄송해요 😢

ddongule commented on src/components/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,27 @@
+    }
+    return this.props.children
+  }
+}
```
eol 확인해주세용!

ddongule commented on src/constants/queryKeys.ts:
```diff
@@ -0,0 +1,24 @@
+  all: () => [QUERY_KEYS.THEMES],
+  products: (themeId: number, params: unknown) =>
+    [QUERY_KEYS.THEMES, themeId, 'products', params] as const,
+};
```
여기도 확인부탁드립니당!

ddongule commented on src/constants/queryKeys.ts:
querykey 👍

SonJH7 replied:
감사합니다!

ddongule commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -0,0 +1,199 @@
+  useWishCountQuery,
+  useWishMutation,
+} from '@/api/product'
+import {
+  Container,
+  ProductImage,
+  Name,
+  Brand,
+  Price,
+  TabBar,
+  TabButton,
+  ContentArea,
+  ReviewCard,
+  ReviewAuthor,
+  ReviewContent,
+  EmptyText,
+  InfoList,
+  InfoItem,
+  StickyBar,
+  LikeSection,
+  LikeButton,
+  LikeCount,
+  OrderButton,
+} from '@/styles/ProductDetailPage.styles'
```
import문이 상당히 길어지네요! 요런 경우에는

```suggestion
import * as Styles from '@/styles/ProductDetailPage.styles'
```
이렇게 한번에 import 후 Styles.Container 처럼 사용할 수도 있어용!

SonJH7 replied:
오 잊고 있었습니다! 감사합니다

SonJH7 replied:
ProductDetailPage.tsx:180 ReferenceError: Container is not defined
    at DetailContent (ProductDetailPage.tsx:105:6)


The above error occurred in the <DetailContent> component.

React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
<DetailContent>		
ProductDetailPage	@	ProductDetailPage.tsx:180
<ProductDetailPage>		
(anonymous)	@	router.tsx:22

이런 에러가 떠서 직접 들고 올 때 뜨지 않아 일단 * 대신 기존 방식으로 PR 올렸습니다 ㅠ

ddongule commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -0,0 +1,199 @@
+  OrderButton,
+} from '@/styles/ProductDetailPage.styles'
+
+function HeartIcon({ filled }: { filled: boolean }) {
+  return (
+    <svg
+      width="20"
+      height="20"
+      viewBox="0 0 24 24"
+      fill={filled ? '#2a3038' : 'none'}
+      stroke="#2a3038"
+      strokeWidth="1.5"
+      strokeLinecap="round"
+      strokeLinejoin="round"
+      style={{ display: 'block' }}
+    >
+      <path d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z" />
+    </svg>
+  )
+}
```
최대한 한 파일에는 하나의 컴포넌트만 있는게 좋아용!
특히 요런 아이콘 같은 경우에는 다른 곳에서도 쓰일 수 있는 가능성이 큰 컴포넌트라 분리해주심 좋을 것 같아요!

SonJH7 replied:
넵 리팩토링 진행하겠습니다!

ddongule commented on src/pages/ProductDetailPage.tsx:
```diff
@@ -0,0 +1,199 @@
+function AnnouncementList({
+  items,
+}: {
+  items: ProductAnnouncementItem[] | Record<string, string> | undefined
```
아이템의 타입은 ProductAnnouncementItem[] 이면 될 것 같은데 혹시 뒤 타입들은 왜 필요한걸까용?

SonJH7 replied:
필요없는 것 같습니다...

ddongule reviewed (changes requested):
안녕하세요 정훈님~! 먼저 리뷰가 늦어지게되어 죄송합니다 ㅠㅠ 
전반적으로 prettier, eol 이슈를 한번 봐주시면 좋을 것 같습니다!
이외에는 소소한 코멘트 남겨두었으니 한번 확인 부탁드립니당! 
고생 많으셨어요! 2단계 머지되고나서 3단계 리뷰도 빠르게 하도록 하겠습니다!
화이팅입니닷!

SonJH7 committed:
chore(prettier): npm run format을 통해 전반적은 prettier, eol 이슈를 해결

SonJH7 committed:
fix(fetchapioptions): method 속성을 httpmethod 유니언 타입으로 제한하여 허용된 http 메서드만 받을 수 있게 처리

SonJH7 committed:
refactor(client): api 응답 본문을 JSON으로 해서하기 전 Context-Type 헤더가 application/json 인지 확인하도록 변경

SonJH7 committed:
refactor(productapi, detailpage): non-null assertion을 제거하고 안전한 기본 값 전달하도록 수정 및 product 페이지 import문 수정

SonJH7 committed:
refactor(productdetail): HeartIcon 컴포넌트 분리 및 공지 리스트 타입 정리

- HeartIcon을 독립 파일로 추출해 재사용성 확보
- ProductDetailPage에서 HeartIcon을 import하여 사용하고, AnnouncementList의 기본 props를 안전하게 처리
- AnnouncementList가 ProductAnnouncementItem[]만 다루도록 타입을 단순화하고, ProductDetail 타입의 announcement 필드를 선택적 배열로 조정

SonJH7 committed:
prettier, eol 세팅

ddongule reviewed (approved):
안녕하세요 정훈님! 리뷰 꼼꼼하게 반영해주신 것 같아 요 PR은 이만 머지하도록 할게요!
고생하셨습니다! 조금만 더 힘을 내봅시다! 으쌰으쌰
